### ISSUE TITLE
[Vote] Register routes directly in Actions

### ISSUE THREAD

**Issue by granitibrahimi** [lorisleiva/laravel-actions#3]
Add a field on each action where routes can be defined and so everything can be setup into the action and no need to create routes or controllers its just create Activity define everything and good to go!

**Comment by lorisleiva** [comment]
Hi, thanks for creating the issue. 👍 

Instead of a field, I was thinking of a function that gives the user more flexibility when defining the routes. Something like that:

```php
class MyAction extends Action
{
    public function routes($router)
    {
        $router->get('/my/action/endpoint', static::class);
    }
}
```

That way they can potentially add multiple endpoints for the same actions and they have access the full router API provided by Laravel (middleware, groups, ect.).

What do you think?

**Comment by granitibrahimi** [comment]
Do you think more then one route is needed for an action?

Im more into having strictly one route per action and I love the idea of router access that is awesome :)

**Comment by lorisleiva** [comment]
Yeah, I would say I have at least one case per application where two routes refer to the same controller's action. As a bonus, users can configure their routes using the API they already know.

Next, I'm going to have to dig into the best ways for discovering those `routes` methods. It assumes that the package checks in a directory that defaults to `app/Actions` but can be configured. I might get some inspiration with the way Laravel auto-discovers policies from 5.8.

If anyone wants to contribute to this task, feel free to send a PR. It might be quicker than waiting for me at the moment.

**Comment by deviouspk** [comment]
Honestly there's a reason why routes are placed in seperate files. I would strongly advise against defining routes on the action class itself. The goal of this package is and should be to glue the components that were previously present in the controller methods together in one class. Making this a superclass where you define everything would be a mistake. Also the only key component that is missing in this package is a way to define a transformer that will be triggered when the action is run from a controller or route. (https://laravel.com/docs/5.8/eloquent-resources)

If you do want to autoregister here's a snippet that should help you:

```
    function getActionFiles()
    {
        try {
            $fileNames = array_diff(scandir("ACTION_FOLDER_PATH"), ['..', '.']);
        } catch (\ErrorException $e) {
            $fileNames = [];
        }

        return $fileNames;
    }
```

```
function getFileNamespace ($filePath) {
	$tokens = token_get_all($filePath);
	$count = count($tokens);
	$i = 0;
	$namespace = '';
	$namespace_ok = false;
	while ($i < $count) {
		$token = $tokens[$i];
		if (is_array($token) && $token[0] === T_NAMESPACE) {
			// Found namespace declaration
			while (++$i < $count) {
				if ($tokens[$i] === ';') {
					$namespace_ok = true;
					$namespace = trim($namespace);
					break;
				}
				$namespace .= is_array($tokens[$i]) ? $tokens[$i][1] : $tokens[$i];
			}
			break;
		}
		$i++;
	}
	if (!$namespace_ok) {
		return null;
	} else {
		return $namespace;
	}
}
```
```
function constructClass($class){
try {
            $reflectionClass = new \ReflectionClass($class);
        } catch (ReflectionException $e) {
            return;
        }
        if ($reflectionClass->isAbstract()) {
            return;
        }

        return $reflectionClass->newInstanceWithoutConstructor();
}
```
```
function loadRoutes(){
        foreach(getActionFiles() as $filePath){
                $actionClass = getFileNamespace($filePath);
                $action = constructClass($actionClass);
                $action->route();
        }
}
```

Hope this help

**Comment by lorisleiva** [comment]
Hi @deviouspk, thanks for your input. That was also my thought when implementing this package, which is why I drew the limit at the routes. However, if it is purely optional and deactivated by default, then I don't really see the harm, do you?

I'd genuinely like more input from users on this feature though. Maybe I could leave this issue to get upvoted for a bit and see if it's a popular demand.

**Comment by deviouspk** [comment]
Yes & no it comes with a cost. You need to discover the files even if you are not using it. Unless you make a config option where you can explicitly turn it off (but this in it's turn increases complexity & confusion). But honestly the people want this feature probably don't have enough software architecture experience to realize this is bad design.

What i would love though as extra feature would be something like this. This can be accomplished with doctrine/annotations (which is present in laravel by default if i'm not mistaken). Tell me what you think ;)

![image](https://user-images.githubusercontent.com/7100315/58661548-6ea8d680-8328-11e9-8eb6-d4989dce619a.png)

![image](https://user-images.githubusercontent.com/7100315/58661645-a879dd00-8328-11e9-8646-d91481eca895.png)

**Comment by lorisleiva** [comment]
No, when I mean deactivated by default I really mean it. I was thinking of making users activate it using a manager class in their route service provider. Something like this:

```php
// Default to finding routes in app/Actions
Actions::loadRoutes();

// Explicitly provide the action directory.
Actions::loadRoutes('my/path');
```

I know what you mean and I personally think it makes sense to keep all routes together like a big dictionary you can refer to but, at the same time, the whole point of actions is to provide a new "unit of life" to your application. A sandbox that contains everything you need for that particular task. So, semantically, I do see the attraction for such a feature.

Regarding `doctrine/annotations`, I'm pretty sure it's not native to Laravel. Are you managing to make it work on your current project? If you'd like to talk more about this, please open a new issue to avoid losing focus. Thanks for your input btw, it's always valuable. 👍

**Comment by deviouspk** [comment]
If you do load the routes don’t forget to cache them with a small TTL. 

Regarding doctrine i think it’s part of a package that laravel uses probably symfony. 
Yes i’ve implemented this in another package and it is working :) i used it to create a dto package (https://github.com/larapie/data-transfer-object). But i’m tired of all the boilerplate that is necessary (dto, requests, controller, service,...). I would love to use a glue class like this ;)

**Comment by deviouspk** [comment]
you are right. i checked & doctrine annotations is not an included as a third party library in laravel. Doesn't really matter though :p What do you think about the idea in general. Declaring public properties to specify what attributes are necessary & validate through the attribute classes (through annotations) or the rules method (maybe merge them) ?

**Comment by ejunker** [comment]
I'm intrigued by this idea. I don't usually like the idea of annotations but I think it could work well for this use case. Also, if you remember at one point Taylor was experimenting with route annotations on controllers but removed it and now it lives in [LaravelCollective/annotations](https://github.com/LaravelCollective/annotations). Also, it looks like PHP 8 will have native support for [annotations/attributes](https://wiki.php.net/rfc/attributes_v2) so it may become more common to use annotations with PHP.

It may have been assumed, but I think it would be important to make sure the routes defined on actions are cached when running `artisan route:cache`

**Comment by lorisleiva** [comment]
@ejunker It's a good idea and I'm not against the idea of supporting annotations. However, it would need to be in addition to supporting a more flexible approach using actual code with an actual Router instance.

I'm not going to overcomplicate this package for now and stick to a `routes` static method but if/when annotations become more popular in the Laravel community I will gladly consider supporting annotations as well.

As I'm writing in this thread, I'd also like to add that this has already been implemented and is available in the `master` branch. I am doing some polishing and re-writing the documentation in an external website before tagging a new release. I'm hoping this can be version `v1.0.0`

**Comment by lorisleiva** [comment]
Released in `v1.0.0`.
